Makefile

```diff
@@ -1,20 +1,28 @@
-
 KERNEL_VERSIONS = 2013.4 2016.4
 
-drivers = intc sarspi uio
+drivers = intc sarspi uio allocated-gpio
 
 default: 2016.4
 
 all: $(KERNEL_VERSIONS)
 
 201%: KVER=$@
+201%:	KREL=$(shell cat $(KDIR)/include/config/kernel.release)
 201%:
 	./version.sh
 	$(foreach dir,$(drivers),$(MAKE) -C $(dir) all;)
-	mkdir -p build/$@
-	$(foreach dir,$(drivers),mv $(dir)/*.ko build/$@;)
-	$(foreach dir,$(drivers),cp $(dir)/*rules build/;)
+	mkdir -p build/udev/
+	cp */*rules build/udev/
+	mkdir -p build/$@/usr/lib/modules/$(KREL)/kernel/imsar
+	rsync -a $(KDIR)/arch/arm/boot/uImage build/$@/
+	rsync -a $(KDIR)/usr/lib/modules/$(KREL)/kernel build/$@/usr/lib/modules/$(KREL)/
+	rsync -a $(KDIR)/usr/lib/modules/$(KREL)/modules.order build/$@/usr/lib/modules/$(KREL)/
+	rsync -a $(KDIR)/usr/lib/modules/$(KREL)/modules.builtin build/$@/usr/lib/modules/$(KREL)/
+	$(foreach dir,$(drivers),mv $(dir)/*.ko build/$@/usr/lib/modules/$(KREL)/kernel/imsar;)
 
-clean:
+clean_all:
 	$(foreach dir,$(drivers),$(MAKE) -C $(dir) clean;)
 	rm -r build
+
+#for KDIR
+include drivers.mk
```

Comment:
When I hit `make clean` I get:
```
[jhancock@watson drivers]$ make clean
make[1]: Entering directory `/home/jenkins/workspace/linux-xlnx/build/2016.4'
/home/jhancock/imsarllc/drivers/Makefile:28: drivers.mk: No such file or directory
make[4]: *** No rule to make target `drivers.mk'.  Stop.
make[3]: *** [_clean_/home/jhancock/imsarllc/drivers] Error 2
make[2]: *** [sub-make] Error 2
make[1]: *** [__sub-make] Error 2
make[1]: Leaving directory `/home/jenkins/workspace/linux-xlnx/build/2016.4'
make: *** [clean] Error 2
```

Reply:
Yes...  at the top level you have to run 'make clean_all'.  I am not sure how else to do that, because there is a clean rule at both the top level Makefile and in drivers.mk.  If they have the same name then Make complains about overriding one or the other.

Reply:
Ok, so the clean rule in the drivers.mk is to be used when building a specific driver I assume?